lib/evmone/eof.cpp

```diff
@@ -90,37 +166,285 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+                return false;
+            rjumpdests.push_back(static_cast<size_t>(jumpdest));
+        }
+        else if (op == OP_RJUMPV)
+        {
+            constexpr auto REL_OFFSET_SIZE = sizeof(int16_t);
+
+            const auto count = container[op_pos + 1];
+
+            const auto post_offset =
+                static_cast<int32_t>(1 + 1 /* count */ + count * REL_OFFSET_SIZE /* tbl */);
+
+            for (size_t k = 0; k < count; ++k)
+            {
+                const auto rel_offset_hi =
+                    container[op_pos + 1 + 1 + static_cast<uint16_t>(k) * REL_OFFSET_SIZE];
+                const auto rel_offset_lo =
+                    container[op_pos + 1 + 2 + static_cast<uint16_t>(k) * REL_OFFSET_SIZE];
+                const auto rel_offset = static_cast<int16_t>((rel_offset_hi << 8) + rel_offset_lo);
+
+                const auto jumpdest = static_cast<int32_t>(i) + post_offset + rel_offset;
+
+                if (jumpdest < 0 || jumpdest >= code_size)
+                    return false;
+
+                rjumpdests.push_back(static_cast<size_t>(jumpdest));
+            }
+
+            const auto rjumpv_imm_size = size_t{1} + count * REL_OFFSET_SIZE;
+            std::fill_n(
+                immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, rjumpv_imm_size, true);
+            i += rjumpv_imm_size;
+            continue;
+        }
+
+        const auto imm_size = instr::traits[op].immediate_size;
+        std::fill_n(immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, imm_size, true);
+        i += imm_size;
+    }
+
+    // Check relative jump destinations against immediate locations.
+    for (const auto rjumpdest : rjumpdests)
+        if (immediate_map[rjumpdest])
+            return false;
+
+    return true;
+}
+
+std::pair<EOFValidationError, int32_t> validate_max_stack_height(
+    bytes_view code, size_t func_index, const std::vector<EOF1TypeHeader>& funcs_in_outs)
+{
+    assert(code.size() > 0);
+
+    // Changing this type to smaller integer (uint16_t) have in mind that it can overflow.
+    std::vector<int32_t> stack_heights = std::vector<int32_t>(code.size(), -1);
+    std::stack<size_t> worklist;
+
+    int32_t stack_height = 0;
+    stack_heights[0] = funcs_in_outs[func_index].inputs_num;
+    worklist.push(0);
+
+    size_t i = 0;
+    Opcode opcode = {};
+    std::vector<size_t> successors;
+    while (!worklist.empty())
+    {
+        i = worklist.top();
+        opcode = static_cast<Opcode>(code[i]);
+        worklist.pop();
+
+        auto stack_height_required = instr::traits[opcode].stack_height_required;
+        auto stack_height_change = instr::traits[opcode].stack_height_change;
+
+        if (opcode == OP_CALLF)
+        {
+            auto fid_hi = code[i + 1];
+            auto fid_lo = code[i + 2];
+            auto fid = static_cast<uint16_t>((fid_hi << 8) | fid_lo);
+
+            if (fid >= funcs_in_outs.size())
```

Comment:
Suggestion: do this check earlier in `validate_instructions`, because as I see it, it is checking the validity of CALLF instruction. Later I would also put there checking immediates of CREATE3 and RETURNCONTRACT.

Reply:
Good idea. It requires extending `validate_intructions` arguments list and pass EOF1TypeHeaders list. I would rather leave a FIXME  comment here and do it after `eof` merge ok?

Reply:
It's ok. In CREATE3 I changed `validate_instruction` to get `EOF1Header`:
https://github.com/ethereum/evmone/pull/553/files#diff-6c68f99275b828a78c2791fee776edf5c210a321c916e8afc58689b1d4d5762dR255-R256